Switch performance tests to Playwright

WunderBart opened the pull request:
## What?

A follow-up to #51084

Closes #51383 
Closes #51379

This PR effectively switches our performance tests to be handled by Playwright. Apart from that, the following improvements have been made:

- Make sure all the tests are consistent in terms of methodology and that the outlying samples (as shown [here](https://github.com/WordPress/gutenberg/pull/51084#issuecomment-1607658262)) are filtered out to improve metrics stability.
- Use the [attach API](https://playwright.dev/docs/api/class-testinfo#test-info-attachments) to handle result files (as suggested by @kevin940726 [here](https://github.com/WordPress/gutenberg/pull/51084#discussion_r1231995414)).
- Make the [performance.js](https://github.com/WordPress/gutenberg/blob/a045f9ec5d8ec9efd186f484ca3fc4d050abb44c/bin/plugin/commands/performance.js) comparison runner result-agnostic (as suggested by @youknowriad [here](https://github.com/WordPress/gutenberg/pull/49238#issuecomment-1551500521)), and move the results to be handled by the dedicated [PerformanceReporter](https://github.com/WordPress/gutenberg/blob/1f8ecd1323cd238eb98447ec56888f5760a29811/test/performance/config/performance-reporter.ts) script.
- Make the PerformanceReporter handle results when running both locally and in CI. Previously, the Reporter was handling only local results which resulted in a discrepancy between CI and local results.
- Make the [performance.js](https://github.com/WordPress/gutenberg/blob/a045f9ec5d8ec9efd186f484ca3fc4d050abb44c/bin/plugin/commands/performance.js) runner suite-agnostic by automatically loading suites from the filesystem.
- Upload the curated results JSON to the artifacts zip as well for easier inspection.

ℹ️ **The [codevitals](https://www.codevitals.run/project/gutenberg) app requires no update as the test results will be produced and uploaded as expected.**

I ran a couple of comparisons of this branch vs. the base branch (round [1](https://github.com/WordPress/gutenberg/pull/52022/commits/d649aacefeb62d343d36a4d6b36a3fa701af5124), [2](https://github.com/WordPress/gutenberg/pull/52022/commits/ae40542df6325a04bfeb24ffe4015d447cc30434) and [3](https://github.com/WordPress/gutenberg/pull/52022/commits/18fe132a027b87d4025f52599ecb3465cad1617a)). The results seem stable and should not disrupt the current codevitals logs:

   | round 1 | round 2 | round 3 |
   | ------------- | ------------- | ------------- |
   | ![](https://github.com/WordPress/gutenberg/assets/1451471/2e0d5ba9-531e-4061-919c-1523accdee40) | ![](https://github.com/WordPress/gutenberg/assets/1451471/ea3e0b53-e5fa-4bd8-bf08-58490688db40) | ![](https://github.com/WordPress/gutenberg/assets/1451471/5b2ff08b-e5b4-4151-8866-c128cabad054) |


## Why?
As a part of the [Playwright migration task](https://github.com/WordPress/gutenberg/issues/38851), but also to improve the overall experience and quality of our performance metrics.


## Testing Instructions

- To run performance tests locally, the command remains unchanged, e.g.:

   ```
   npm run test:performance -- post-editor --headed
   ```
   
   You should see different reporter output though, as it was updated to be consistent with the format we use in CI. Also, missing front-end perf results have been added:
   
   | Before | After |
   | ------------- | ------------- |
   | ![](https://user-images.githubusercontent.com/1451471/245774797-5b0b4377-29ad-47fa-bcfa-735f5645e899.png) | ![](https://github.com/WordPress/gutenberg/assets/1451471/62f5c54d-20e8-4892-9315-58eb4f410069) |

- To run branch performance comparison tests locally (via the CLI tool), the command also remains unchanged as well as the output, e.g.:

   ```
   ./bin/plugin/cli.js perf my-branch trunk --tests-branch my-tests-branch
   ```
   
   You can check the performance GH action output for this PR to verify the above as well.

WunderBart committed:
Run with Playwright in CI

WunderBart committed:
Print stdout on error

WunderBart committed:
Run build:packages to use e2e utils

WunderBart committed:
Extend timeout for the canvas spinner

WunderBart committed:
Add missing front end perf report

WunderBart committed:
Remove obsolete step
The first character is excluded from the results later on, so let's not double-exclude.

WunderBart committed:
Make the typing testing consistent

WunderBart committed:
Improve post editor perf testing consistency
- define fixed time to wait until browser becomes idle instead random values
- make each test record 10 samples + 1 throwaway to address initial metric spikes
- improve overall code consistency

WunderBart committed:
Do not report slow tests

WunderBart committed:
Use the same methodology for measuring post and site editors metrics

WunderBart committed:
Make the results table transformer more readable

WunderBart committed:
Streamline the results processing p.1

WunderBart committed:
Improve visual step separation a tad :)

WunderBart committed:
First step to separate metrics from the runner

WunderBart committed:
Define the results file suffix

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/playwright-performance-tests-p2

WunderBart committed:
cosmetic

WunderBart committed:
Fix matching pattern typo

WunderBart committed:
Make performance.js comparison runner results-agnostic

WunderBart committed:
Fix results handling

WunderBart committed:
Use Playwright's attachments API to handle results

WunderBart committed:
Upload raw results as well

WunderBart committed:
Be consistent with calc utils

WunderBart committed:
Streamline results handling

WunderBart committed:
Use the same sampling convention for front-end metrics

WunderBart committed:
Fix the metrics output in perf runner

WunderBart committed:
tmp: compare performance with base branch

WunderBart committed:
tmp: compare performance with base branch (run 2)

WunderBart commented on test/performance/specs/site-editor.spec.js:
```diff
@@ -14,13 +14,14 @@ const path = require( 'path' );
 	loadBlocksFromHtml,
 } = require( '../utils' );
 
+const BROWSER_IDLE_WAIT = 1000;
```
ℹ️ Based on https://github.com/WordPress/gutenberg/issues/51383#issuecomment-1613460429

kevin940726 replied:
Let's move this to an inline comment?

WunderBart committed:
tmp: compare performance with base branch (run 3)

WunderBart committed:
Revert "tmp: compare performance with base branch"

This reverts commit d649aacefeb62d343d36a4d6b36a3fa701af5124.

WunderBart committed:
Remove unused code and do the switch

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/playwright-performance-tests-p2

dmsnell commented on test/performance/specs/site-editor.spec.js:
```diff
@@ -168,14 +175,8 @@ test.describe( 'Site Editor Performance', () => {
-		await firstParagraph.click();
-		await pageUtils.pressKeys( 'primary+a' );
-		await page.keyboard.press( 'ArrowRight' );
-		await page.keyboard.press( 'Enter' );
```
I assume there's a reason these are all gone here, but are we certain they didn't help reset some state which was helpful for the tests?

WunderBart replied:
The reason for those changes is to bring consistency in methodology. To make these changes easier to follow, I've made some _before_ and _after_ recordings and put them together in [this comment](https://github.com/WordPress/gutenberg/pull/52022#issuecomment-1651225241). Hope it helps! 
 
BTW, this test actually needed a small tweak that I pushed [here](https://github.com/WordPress/gutenberg/pull/52022/commits/2c305da1b9353890b15149490d0c2c0e165b4037). The change was intended but still not working as expected, so nice catch! 😄

WunderBart replied:
Added a comment in https://github.com/WordPress/gutenberg/pull/52022/commits/894896bb3ac81ac4ed0f43b762708e0b87b69a4d

dmsnell commented on test/performance/specs/front-end-block-theme.spec.js:
```diff
@@ -19,19 +14,27 @@ test.describe( 'Front End Performance', () => {
+
 			// Go to the base URL.
-			await page.goto( '/', { waitUntil: 'networkidle' } );
+			await testPage.goto( '/', { waitUntil: 'networkidle' } );
```
> DISCOURAGED consider operation to be finished when there are no network connections for at least 500 ms. Don't use this method for testing, rely on web assertions to assess readiness instead.

I guess we still want this despite the docs discouraging it?

WunderBart replied:
Yeah, I'm aware but unsure what to replace it with. Now that I'm looking at it, maybe we can just remove it as in [this example](https://www.checklyhq.com/learn/headless/basics-performance/#:~:text=Largest%20Contentful%20Paint%20API%20(largest%2Dcontentful%2Dpaint))?

dmsnell reviewed (commented):
Just a couple quick questions, but it looks like you've been thorough here. I don't know if there are things otherwise I can review without looking at it in production, but the things are questions about integrating with existing services like the codehealth app, and I think you have taken care of that already.

So, well done.

kevin940726 commented on packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts:
```diff
@@ -66,5 +66,5 @@ export async function visitSiteEditor(
 	await this.page
 		.locator( '.edit-site-canvas-spinner' )
-		.waitFor( { state: 'hidden' } );
+		.waitFor( { state: 'hidden', timeout: 60_000 } );
```
Why do we need this here? If it applies to all the tests then maybe we should add a comment above explaining why 😅 .

WunderBart replied:
This was needed because, for performance tests, we load [an enormous post fixture](https://github.com/WordPress/gutenberg/blob/refactor/playwright-performance-tests-p2/test/performance/assets/large-post.html) to test against, which often takes more than the default 10 seconds to load.

I'm not sure the rationale for using a post big enough to freeze the browser for a bit, but it's probably something to be discussed in a separate PR? /cc @youknowriad

dmsnell replied:
when I've tested with the large post in my browser it opens reasonably fast, but if I open the profiling tools I've seen it freeze for more than a minute.

I'm inferring that Chrome is spending a lot of time instrumenting the thousands of listeners and hooks when we turn of tracing, which probably happens here as well.

so this may not be realistic from a user perspective, but the wait may still occur when testing and profiling

kevin940726 replied:
Yeah, profiling adds a lot of overhead. We could assume that the overhead added shares averagely across all code and can be seen as a fair emulated slowdown of the CPU, but I don't know if that's true 😅 .

dmsnell replied:
it's weird. I don't understand it, but it doesn't seem to appear in the metrics. I suspect that the profiling mechanisms don't start until after this initialization. the page isn't interactive while it's starting up, but then once it finishes things are suddenly fast and fine.

so it may not be the case that Chrome reports this big delay, but it still delays the tests. based on some recent profiling work I'm wondering if we might have code paths that are to blame. there are so many listeners involved and it doesn't seem reasonable that we should have so many. maybe Chrome's profiler simply chokes at tens of thousands of listeners (or maybe this is a normal number 🤷‍♂️ I really don't know).

WunderBart committed:
Fix site editor typing spec

WunderBart commented:
For further inspection, here's a comparison of before and after the migration + proposed fixes for metric consistency. The most important changes are:

- For the typing tests, we now consistently wait between each keystroke as suggested in [this comment](https://github.com/WordPress/gutenberg/issues/51383#issuecomment-1584893744). This has been applied to all typing tests in both post and site editor suites. The wait value between the keystrokes has been based on measurements presented in [this comment](https://github.com/WordPress/gutenberg/issues/51383#issuecomment-1613460429).
- For the loading tests, instead of reloading the same page for each sample, we create a new page in a separate browser context.
- For the UI metrics (e.g., `inserterOpen` or `listViewOpen`), the first sample has been filtered out as an outlier. This was based on measurements presented in [this comment](https://github.com/WordPress/gutenberg/pull/51084#issuecomment-1607658262:~:text=firstBlock-,UI%20performance,-Metric).



| Test                                                        | Before (Puppeteer)                                                                                               | After (Playwright)                                                                                               |
|-------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------|
| Front End › Block Theme › Report TTFB, LCP, and LCP-TTFB    | <video src="https://github.com/WordPress/gutenberg/assets/1451471/4c63b62c-9b55-4671-8389-bd292d8c9fd1"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/43ab4dd3-25b1-4733-b98c-781feb3dbc19"></video> |
| Front End › Classic Theme › Measure TTFB, LCP, and LCP-TTFB | <video src="https://github.com/WordPress/gutenberg/assets/1451471/bad3e989-3978-4075-93b1-88c820652839"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/b7b3ee13-8821-4ff7-89a2-524519e98f2b"></video> |
| Post Editor › Loading                                       | <video src="https://github.com/WordPress/gutenberg/assets/1451471/a3deb029-ddac-4046-8e83-b75cb7608e83"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/d4e7aea8-34b6-496d-bfe1-35a9b683456e"></video> |
| Post Editor › Typing                                        | <video src="https://github.com/WordPress/gutenberg/assets/1451471/06c63356-eaf5-4a04-a921-8c94d7209fb1"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/21e162b9-a9c1-4e29-8589-eaa557bc911c"></video> |
| Post Editor › Typing within containers                      | <video src="https://github.com/WordPress/gutenberg/assets/1451471/5462477e-a933-4b26-bc40-424a6fc10889"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/f27170b9-b0f1-41a0-818e-f8d5a8caa0ff"></video> |
| Post Editor › Selecting blocks                              | <video src="https://github.com/WordPress/gutenberg/assets/1451471/81025ead-8b00-414f-9829-eaa77d55c6aa"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/c6fe1d66-3285-4ec4-89d3-9c44de495583"></video> |
| Post Editor › Opening persistent list view                  | <video src="https://github.com/WordPress/gutenberg/assets/1451471/b29f6c88-33af-45fd-bc42-4721ec0f3265"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/edb4a025-fc7e-4462-8924-87ddf69ff988"></video> |
| Post Editor › Opening the inserter                          | <video src="https://github.com/WordPress/gutenberg/assets/1451471/5e542cb1-34f1-409d-a641-55efe2d5bb0f"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/42d0bdb0-c28f-4914-b91e-d96ee9aebdcc"></video> |
| Post Editor › Searching the inserter                        | <video src="https://github.com/WordPress/gutenberg/assets/1451471/7d430121-e9d6-4171-8e6a-d18653264c3f"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/097854e3-301a-431a-a062-a50942c34815"></video> |
| Post Editor › Hovering Inserter Items                       | <video src="https://github.com/WordPress/gutenberg/assets/1451471/71ed9046-f092-459f-9232-b97594d9fbb4"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/2164a013-dbdb-414d-beb9-e8a9c6a11dbd"></video> |
| Site Editor › Loading                                       | <video src="https://github.com/WordPress/gutenberg/assets/1451471/5a8a2f71-84a0-45be-8e80-c0f3d3f0877e"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/15a827fd-c312-451c-a41e-14a65a2fc4da"></video> |
| Site Editor › Typing                                        | <video src="https://github.com/WordPress/gutenberg/assets/1451471/05f781bd-406f-469f-896a-2405b6f7ec5d"></video> | <video src="https://github.com/WordPress/gutenberg/assets/1451471/dc337d4d-8580-421d-8491-5644984d3f29"></video> |

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/playwright-performance-tests-p2

kevin940726 commented on bin/plugin/lib/utils.js:
```diff
@@ -120,10 +121,32 @@ function getRandomTemporaryPath() {
+		return [];
+	}
+
+	const files = fs.readdirSync( dir );
+	const filelist = files
+		.map( ( file ) => path.join( dir, file ) ) // map to full file path
+		.filter( ( file ) => fs.lstatSync( file ).isFile() ); // filter out subdirectories
```
Nit: We could also pass `withFileTypes: true` to do it with one call.

```js
fs.readdirSync( dir, { withFileTypes: true } )
  .filter( dirent => dirent.isFile() )
  .map( path.join( dir, dirent.name ) )
```

kevin940726 commented on test/performance/specs/front-end-block-theme.spec.js:
```diff
@@ -19,19 +14,27 @@ test.describe( 'Front End Performance', () => {
+		const rounds = samples + throwaway;
+
+		for ( let i = 0; i < rounds; i++ ) {
+			const testPage = await browser.newPage( { storageState: {} } );
```
It's [recommended](https://playwright.dev/docs/api/class-browser#browser-new-page) to spawn a new `context` than a new page directly so that the lifecycle of the context can be better controlled.

kevin940726 commented on test/performance/specs/site-editor.spec.js:
```diff
@@ -185,22 +192,31 @@ test.describe( 'Site Editor Performance', () => {
+		} );
 
 		// Stop tracing and save results.
 		await browser.stopTracing();
```
The tracing API also returns the [buffer](https://playwright.dev/docs/api/class-browser#browser-stop-tracing). So maybe we don't need to store this one-off trace to a file on disk after all and can just process them in memory.

WunderBart replied:
I've noticed that the trace file is not actually attached to the failure artifacts (in case the test fails mid-tracing), so it should be OK to only use the buffer instead.

kevin940726 commented on test/performance/utils.js:
```diff
@@ -2,12 +2,56 @@
+import { existsSync, readFileSync, unlinkSync } from 'fs';
+
+export function sum( array ) {
+	if ( ! array || ! array.length ) return undefined;
```
Nit: For testing I think it might be better to just throw an error if we're passing unexpected arguments.

WunderBart replied:
I think the error throwing should happen one step earlier (where we collect results data, for example, via `expect()`). Before that, though, I wanted to migrate the tests to TypeScript (in a separate PR) to get a better feel for how to handle exceptions.

For these calc functions (sum, median, average, etc.) I think it's OK not to throw, but it's still important to stay consistent in how we handle invalid arguments (`return undefined`). Having said that, I'm aware this implementation is not super-defensive, but I don't think it's necessary at this point.

kevin940726 replied:
If we were using TypeScript, we wouldn't have to guard the inputs during runtime since the error would be caught statically. 👍

kevin940726 reviewed (approved):
Nice! Left some suggestions but all minor.

I think this looks good but given that it might have a bigger impact I don't object to having more eyes on this before merging 💯 .

WunderBart committed:
Add inline comment

WunderBart committed:
Expain why we need larger timeout for site editor visit util

WunderBart committed:
Use dirent for getting files from dir

WunderBart committed:
Do not use the browser.newPage shortcut
https://playwright.dev/docs/api/class-browser#browser-new-page

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/playwright-performance-tests-p2

WunderBart committed:
Use trace buffer instead of tmp file

swissspidy commented on test/performance/specs/front-end-block-theme.spec.js:
```diff
@@ -19,19 +14,30 @@ test.describe( 'Front End Performance', () => {
+		const rounds = samples + throwaway;
+
+		for ( let i = 0; i < rounds; i++ ) {
+			const tmpContext = await browser.newContext( {
+				storageState: {}, // User will be logged out.
+			} );
+			const tmpPage = await tmpContext.newPage();
```
Interesting nuance to test as a logged out user 👍 

Is it necessary to create the new context within the loop or could that be done only once before to speed things up? Or is that to ensure empty caches?

WunderBart replied:
By always starting from a fresh state, we aim to reduce the number of variables and ensure that each test run is as similar as possible to the others. The variables can be caching, state side-effects, memory management, etc.

dmsnell replied:
@swissspidy we did discover that re-using a session, even after telling it to clear cache, was biasing the performance results above the difference caused by the vast majority of performance-impacting changes.

wherever possible in the performance tests, we should create a new page, new tab, new session, to avoid this.

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/playwright-performance-tests-p2

WunderBart committed:
Fix 'unterminated string' error

WunderBart committed:
Fix 'unterminated string' error p.2

WunderBart committed:
Make sure sanitized branch name is used for reading results json

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/playwright-performance-tests-p2

WunderBart committed:
Prevent logging in via test.use() instead of creating new context
Each test is run with a fresh context by default, so we can just set the storageState via test.use instead.
